### ISSUE TITLE
Upgrade to websocket after handshake/authentication

### ISSUE THREAD

**Issue by utkarshkap** [kerryjiang/SuperSocket#288]
Kerry, I want to authorize user before upgrade connection to websocket is it possible?

For eg: wss://test.myserver.com:8080?uid=abc&pwd=xyz

Any help in this regard is appreciated.

Utkarsh

**Comment by kerryjiang** [comment]
It is a good point. I will consider exposing an API for handshake.

**Comment by utkarshkap** [comment]
Kerry, 

Great to hear from you! Thank you for considering.

Also I tried to upgrade to Tls1.3 but it doesn't seems to be working however Tls1.2 is working fine for now.

Thank you!

**Comment by kerryjiang** [comment]
Added this feature by HandshakeOptions:
https://github.com/kerryjiang/SuperSocket/commit/473e741b1d5cf5f6a50287bb7c5c3904481a4edb

**Comment by kerryjiang** [comment]
@utkarshkap does it work for you.

**Comment by utkarshkap** [comment]
Not tested sorry, will test asap. also strangely I am not able to update Supersocket.Websocket 2.0 from NuGet from VisualStudio 2017, could it be due to Preview tag?

One more thing, I use pub/sub model to send streaming stock quotes to thousands of users over Supersocket websocket, each user gets 100+ messages a second, and I am getting **frequent sending attempt timeout**, can sendqueuesize increase will help? Basically it delays stock quotes for other connected users! I am stuck up

**Comment by kerryjiang** [comment]
"frequent sending attempt timeout"? Which version do you use right now?

The enhancement I mentioned is for SuperSocket 2.0

**Comment by utkarshkap** [comment]
From VS I can only use **1.6.6.1 Supersocket.Websocket** last published on 19-05-2016 ! 

I think pre release versions are not available right from VS. This is a prerelease version of SuperSocket.WebSocket.

**Comment by utkarshkap** [comment]
Even I tried command line
**Install-Package SuperSocket.WebSocket -Version 2.0.0-preview3-03** , but it keeps on saying 

You are trying to install this package into a project that targets '.NETFramework,Version=v4.5', but the package does not contain any assembly references or content files that are compatible with 
that framework. For more information, contact the package author.

**Comment by kerryjiang** [comment]
You have to use 1.6.6.1.
The latest preview versions are for .NET standard.

**Comment by utkarshkap** [comment]
Oh.. any version of Supersocket 2.0 compatible with .Net framework 4.6.1 or .NET standard 2.0?

**Comment by kerryjiang** [comment]
Sorry,  no. You have to move forward to .NET Core completely.

**Comment by utkarshkap** [comment]
.NET Core 3.0 only right?

**Comment by utkarshkap** [comment]
Kerry,

Supersocket library on Nuget shows old version 1.6.6.1(based on .NET Frameworks 4.6 etc) while Supersocket.Websocket shows newer versions 2.0 (on .NET Core). I upgrade to VS 19 with .NET core 3.1, which gave me access to Supersocket.Websocket newer from Nuget while Supersocket is old based on .NET Framework 4.6. I dont know what to do now!

Also, is there any example to build Websocket server with newer Supersocket 2.0 library? 

Any help in this regard is appreciated :)

Thanks

**Comment by kerryjiang** [comment]
Yes, it is the document:
https://docs.supersocket.net/v2-0/en-US/WebSocket-Server

**Comment by utkarshkap** [comment]
Kerry, thanks you for sharing link, I will be starting over with .NET Core 3.0 or above soon.
However, I also wonder how I can achieve two things:
First, multi-casting to selected sessions/users or all broadcasting to all sessions possible, any method?
Second, can I compress data like HTTP gzip compression like the one built in IIS?
Please help.


    On Friday, 1 May, 2020, 10:11:56 am IST, Kerry Jiang <notifications@github.com> wrote:  
 
 


Yes, it is the document:
https://docs.supersocket.net/v2-0/en-US/WebSocket-Server

**Comment by utkarshkap** [comment]
Pl check below things if you have time:

First, multi-casting to selected sessions/users or all broadcasting to all sessions possible, any method?
Second, can I compress data like HTTP gzip compression like the one built in IIS?
Please help.

Sec Header : permessage-deflate is not accepted by SuperSocket, so is it possible to have some compression ?

**Comment by kerryjiang** [comment]
1) it is doable, just fetch sessions you want from session container and then send data with session to the clients;
https://github.com/kerryjiang/SuperSocket/tree/master/samples/WebSocketPushServer
2) permessage-deflate is not supported for now, but I will keep it in the plan;

**Comment by utkarshkap** [comment]
Kerry,

Thanks for your reply!

But I wonder, how I can send/push content to selected subscribers(sessions) instead of all sessions without loop.

Pl help with example if any as I am working with pub/sub model.

Regards

**Comment by kerryjiang** [comment]
We started to support permessage-deflate from beta 5 just now.

**Comment by utkarshkap** [comment]
I appreciate your consideration to address websocket compression
We will test shortly

Sent from Yahoo Mail on Android 
 
  On Mon, 10 Aug 2020 at 3:40 AM, Kerry Jiang<notifications@github.com> wrote:   


We started to support permessage-deflate from beta 5 just now.

**Comment by utkarshkap** [comment]
Kerry,
Before we started migrating to SuperSocket 2.0 we are having following issue with current 1.6 WebSocket Service 

Issue1: 
Listener is stopping without throwing error!
2020-08-28 18:06:13,361 DEBUG Listener (0.0.0.0:8082) was stoppped
2020-08-28 18:06:13,361 INFO The server instance WebSocketService has been stopped!
2020-08-28 18:06:13,362 INFO Session: 4cc3d9b4-d505-45c6-9722-8e683d533f33/34.68.95.93:54956
This session was closed for ServerShutdown!
and so all users are disconnected henceforth...

Issue2:
Unexpected error
Caller: OnBeginInitStream, file path: d:\WorkShop\SuperSocket\v1.6\SocketEngine\AsyncStreamSocketSession.cs, line number: 245
System.IO.IOException: Authentication failed because the remote party has closed the transport stream.
   at System.Net.Security.SslState.InternalEndProcessAuthentication(LazyAsyncResult lazyResult)
   at System.Net.Security.SslState.EndProcessAuthentication(IAsyncResult result)
   at System.Net.Security.SslStream.EndAuthenticateAsServer(IAsyncResult asyncResult)
   at SuperSocket.SocketEngine.AsyncStreamSocketSession.OnBeginInitStream(IAsyncResult result, Boolean connect)

Any help or possible causes of above issues ?

Regards
